### PR TITLE
Gemma4: batched prefill, prefill device trace, and multi-user decode (#44955 and #44957)

### PR DESCRIPTION
**Pull request by dwadhwaniTT**
### PR Category
Feature

## What this PR does 

Gemma4 could already run **one user at a time**. This PR adds the plumbing to run **many users in one prefill pass** when their prompts are the same length, using **paged attention** to keep each user's KV cache separate.

It also adds **prefill device trace** for MoE models (26B-A4B, 31B): compile once, replay many times — but only where it is safe and useful (short/medium ISL, not huge batched token counts).

Finally, it fixes **multi-user decode** performance and correctness (on-device embeddings, faster RMSNorm, better attention head concat) so batch-32 prefill + decode demos actually work end-to-end.

**Important expectation:** For **small batches and short ISLs**, **sequential** prefill (one user at a time) is often **faster** than batched prefill today. Batched prefill is still valuable at larger scale (e.g. 26B on 1×8). The new perf test logs both modes so we can track improvements over time.

---

---

## Issues and acceptance criteria

| Issue | Title (summary) | What “done” means in this PR | What is explicitly out of scope / deferred |
|-------|-----------------|------------------------------|---------------------------------------------|
| **#44955** | Batched prefill | Multi-user batched prefill in `Gemma4Generator`, 128k token chunking, batch-32 demo tests, batched-vs-sequential perf logging | **+50% batched speedup** acceptance criterion (26B-A4B / 1×8 / 32×4096) — **not met on 31B 1×4**; perf test logs only unless `GEMMA4_PREFILL_PERF_ASSERT=1` |
| **#44957** | Prefill device trace | Trace policy (ISL ≤4096, batch×kernel &lt;32k, PLI off), batched trace warmup sweep, parity/perf/Tracy tests, scoped single-bucket warmup in parity tests | Trace at ISL &gt;4096 or ≥32k virtual tokens (disabled by policy); batch-32×2048/4096 on 31B 1×4 (DRAM xfail) |
| **#44962** | Parent tracking | Delivers the batched-prefill and trace workstreams above | Full vLLM production rollout beyond existing parity harness |

---

---

## Features added

1. **Batched prefill forward path** — concatenate equal-length user prompts into one kernel; fill KV per user via paged attention.
2. **128k batched-prefill ceiling with chunking** — when `padded_batch × kernel_len ≥ 128k`, split into smaller user chunks automatically.
3. **Batch-32 demo and integration tests** — pytest coverage for prefill-only and prefill+decode at multiple ISL buckets.
4. **Prefill device trace (MoE only)** — capture/replay with explicit limits; PLI models (E2B/E4B) disable prefill trace.
5. **Trace validation suite** — eager vs traced vs HuggingFace parity, capture-vs-replay perf, Tracy CSV workflow.
6. **Multi-user decode fixes** — device-side token embed, sharded RMSNorm, multi-core decode head concat, per-user RoPE.
7. **Batched prefill perf test** — compares aggregate tok/s for batched vs sequential on configurable `(batch, ISL)`.
8. **Stable pytest markers** — `gemma4_batched_prefill`, `gemma4_prefill_trace`, `gemma4_hf_direct_parity`.

---

## Per-file breakdown

Every file touched under `models/demos/gemma4/` in this PR, what changed, and which scope it serves.

#### `tt/generator.py`
**What:** Extends `Gemma4Generator` with batched-prefill chunking at the 128k ceiling, prefill trace policy hooks, and shared warmup via `generator_trace.py`. Removes the old PLI trace capture-then-clear workaround.  
**Why:** `#44955` — entry point for multi-user prefill through the shared `tt_transformers` Generator batched path. `#44957` — gates trace on batch×ISL before calling `super().prefill_forward_text()`.

#### `tt/generator_trace.py` *(new)*
**What:** Central trace policy: ISL buckets `{128,512,1024,2048,4096}`, max ISL 4096, max 32k batched virtual tokens, PLI guard, full batch×ISL trace warmup sweep, and `warmup_gemma4_prefill_bucket()` for single-bucket test warmup.  
**Why:** `#44957` — one place to enforce “when trace is allowed” and avoid OOM from warming every batch×ISL combo in parity tests.

#### `tt/generator_vllm.py`
**What:** Wires the same trace policy and MoE prefill warmup into the vLLM hybrid generator; applies trace gating in `prefill_forward_text`.  
**Why:** `#44957` — vLLM path must behave like standalone generator for trace capture on MoE models.

#### `tt/model.py`
**What:** Batched prefill through the model stack: passes `batch_size` / `user_id`, flattens embeddings for batched input, returns hidden states (not logits) for batched prefill so Generator applies norm+lm_head per user, ROW_MAJOR 2D RoPE caches for decode embedding lookup, device-side decode embedding in `prepare_decode_inputs_host`, `extract_last_tokens_batched_prefill` / `_apply_norm_and_lm_head` for batched last-token logits.  
**Why:** `#44955` — core forward for concatenated multi-user prefill. `#44952` / decode — multi-user decode staging and batched last-token logit extraction.

#### `tt/model_config.py`
**What:** Expands `get_warmup_prefill_supported_seq_lens()` to include 1024, 2048, 4096.  
**Why:** `#44957` — compile/warm 4k prefill kernels needed for trace and long ISL demo buckets.

#### `tt/layer.py`
**What:** Reshapes layer norm output to `[batch, 1, seq_per_user, hidden]` before attention when `batch_size > 1` in prefill.  
**Why:** `#44955` — attention SDPA expects per-user batch dimension in batched prefill.

#### `tt/rms_norm.py`
**What:** Adds width-sharded RMSNorm fast path for decode-shaped tensors (lazy-built, cached per hidden width).  
**Why:** `#44952` / decode perf — cuts single-token norm latency on large models (31B decode path).

#### `tt/attention/prefill.py`
**What:** Adds multi-user batched prefill path: flatten/reshape hidden states, per-user paged KV fill loop, batched SDPA, reshape back to concatenated sequence.  
**Why:** `#44955` — attention is the main kernel that must understand multiple users in one forward.

#### `tt/attention/decode.py`
**What:** Uses new per-user decode RoPE (`apply_rope_decode_peruser`) and multi-core `nlp_concat_heads_decode` path.  
**Why:** `#44952` — correct and faster multi-user decode after batched prefill.

#### `tt/attention/operations.py`
**What:** Batched-aware Q/K per-head RMSNorm; `apply_rope_decode_peruser`; decode `concat_heads` via height-sharded + `nlp_concat_heads_decode`.  
**Why:** `#44955` (batched prefill norms) and decode perf/correctness for batch &gt; 1.

#### `tt/attention/__init__.py`
**What:** Threads `batch_size` and `user_id` from layer into prefill/decode dispatch.  
**Why:** `#44955` — plumbing to reach batched prefill implementation.

---

#### `demo/text_demo.py`
**What:** Batch-32 (configurable via `GEMMA4_BATCH_DEMO_SIZE`) prefill and prefill+decode demos; helpers for page tables, prompt loading, batch generation; DRAM xfail for `32×{2048,4096}` on 31B 1×4; 128k ceiling documentation test; marker `gemma4_batched_prefill`.  
**Why:** `#44955` / `#44952` — runnable demo and integration tests proving batched prefill + decode work; documents hardware limits honestly.

---

#### `tests/unit/test_batched_prefill_perf.py` *(new)*
**What:** Times batched vs sequential prefill on the same tokens; logs aggregate tok/s; optional speedup assert via env vars. Default mesh gate: 1×4.  
**Why:** `#44955` — measures batched prefill benefit (or lack thereof); does **not** enforce +50% by default.

#### `tests/unit/test_prefill_trace_parity.py` *(new)*
**What:** Matrix: ISL `{128,512,1024,2048,4096}` × batch `{1,2,4,8,16,32}`. Compares traced vs eager Generator, direct model vs HF, traced vs HF when prompt fills bucket. Uses **single-bucket warmup** only (avoids full trace matrix OOM). Marker `gemma4_prefill_trace`.  
**Why:** `#44957` — proves trace does not break numerics; scoped warmup fixes batch-16×1024 OOM from full sweep.

#### `tests/unit/test_prefill_trace_perf.py` *(new)*
**What:** Capture once + replay once timing for one (batch, ISL) under plain pytest.  
**Why:** `#44957` — validates trace amortization without Tracy overhead.

#### `tests/unit/test_prefill_trace_tracy_csv.py` *(new)*
**What:** Full Tracy session with signposted measured replay for `ops_perf_results_*.csv`.  
**Why:** `#44957` — ops perf CSV workflow for kernel optimization.

#### `tests/unit/tracy_prefill_common.py` *(new)*
**What:** Shared fixtures, generator load, capture/replay helpers, Tracy signpost context manager.  
**Why:** `#44957` — DRY support for perf + Tracy tests.

---

#### `tests/conftest.py`
**What:** Registers pytest markers: `gemma4_hf_direct_parity`, `gemma4_batched_prefill`, `gemma4_prefill_trace`.  
**Why:** CI filtering and readable test taxonomy for both workstreams.

#### `tests/unit/test_model.py`
**What:** Adds `test_full_model_decode` (decode-step PCC vs HF); per-token PCC logging in `test_full_model`; marker `gemma4_hf_direct_parity`.  
**Why:** Regression guard for decode path used after batched prefill; not strictly #44955/57 but supports batch-32 decode confidence.

#### `tests/unit/test_vllm_parity.py`
**What:** Decode step inputs use host embeddings (`compute_host_embeddings`) instead of raw token IDs, matching production decode path.  
**Why:** Keeps vLLM parity harness aligned with on-device embedding decode changes in `model.py`.

#### `tests/test_factory.py`
**What:** Adds 8192 to `PREFILL_BUCKETS`; case-insensitive PCC threshold lookup for HF model path casing.  
**Why:** Long-context attention tests; `#44957` parity thresholds resolve correctly for `google/gemma-4-31B-it` vs `gemma-4-31B-it` keys.

#### `tests/pcc_thresholds.json`
**What:** Adds 8192 attention thresholds for 31B 1×4; raises `test_full_model` 1×4 threshold; adds per-node thresholds for `test_prefill_trace_eager_hf_parity` on 31B 1×4.  
**Why:** `#44957` — parity test passes with model-specific PCC floors instead of generic 0.99 default everywhere.

---

#### `README.md`
**What:** Documents multi-user batched demo tests, prefill trace limits, and `gemma4_batched_prefill` marker.  
**Why:** Onboarding — explains what the new tests do without internal jargon.

---

### Batched prefill demo (#44955)
```bash
export HF_MODEL=google/gemma-4-31B-it TT_ARCH=blackhole
pytest models/demos/gemma4/demo/text_demo.py -k "test_demo_batch_prefill" -m gemma4_batched_prefill -v
pytest models/demos/gemma4/demo/text_demo.py -k "test_demo_batch_32" -m gemma4_batched_prefill -v
```

### Prefill trace parity (#44957)
```bash
pytest models/demos/gemma4/tests/unit/test_prefill_trace_parity.py -m gemma4_prefill_trace -k "1x4" -v --timeout=14400
```

### Batched vs sequential perf (logging only)
```bash
export HF_MODEL=google/gemma-4-31B-it TT_ARCH=blackhole
pytest models/demos/gemma4/tests/unit/test_batched_prefill_perf.py -k "1x4" -m gemma4_batched_prefill -v -s
```

### HF direct parity
```bash
pytest models/demos/gemma4/tests/unit/test_model.py -m gemma4_hf_direct_parity -k "1x4" -v
```

---

## Known limitations

- **31B 1×4, batch 32, ISL 2048/4096:** expected xfail (DRAM — ~64k batched activation budget).
- **Prefill trace:** disabled when ISL &gt; 4096 or batch×kernel ≥ 32k; disabled entirely on PLI models.
- **Batched prefill speedup AC (+50%):** not demonstrated on 31B 1×4; sequential often wins at smaller shapes. Long-term target remains 26B-A4B / 1×8 / 32×4096.

---

## Test plan checklist

- [x] `pytest models/demos/gemma4/demo/text_demo.py -m gemma4_batched_prefill -k prefill_128`
- [x] `pytest models/demos/gemma4/tests/unit/test_prefill_trace_parity.py -m gemma4_prefill_trace -k "batch1-prefill_128"`
- [x] `pytest models/demos/gemma4/tests/unit/test_prefill_trace_parity.py -m gemma4_prefill_trace -k "batch16-prefill_1024-1x4"`
- [x] `pytest models/demos/gemma4/tests/unit/test_batched_prefill_perf.py -m gemma4_batched_prefill -k 1x4 -v -s`
- [x] `pytest models/demos/gemma4/tests/unit/test_model.py -m gemma4_hf_direct_parity -k test_full_model_decode`

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:dwadhwaniTT/44952-batch32-demo)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:dwadhwaniTT/44952-batch32-demo)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml?query=branch:dwadhwaniTT/44952-batch32-demo)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:dwadhwaniTT/44952-batch32-demo)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:dwadhwaniTT/44952-batch32-demo)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:dwadhwaniTT/44952-batch32-demo)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:dwadhwaniTT/44952-batch32-demo)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:dwadhwaniTT/44952-batch32-demo)

### CODE REVIEW

**Comment by dwadhwaniTT** [comment]
Running CI

**Comment by dwadhwaniTT** [comment]
Re-running CI after copilot changes.

**Review by mtairum — APPROVED** [review]

**Comment by dwadhwaniTT** [comment]
Resolving merge conflicts

Edit : Done

**Review by williamlyTT — APPROVED** [review]

**Review by nostojicTT — APPROVED** [review]

File: models/common/warmup/warmup_utils.py [hunk 1]
```diff
@@ -87,10 +87,14 @@ def warmup_model_decode(
         can_sample_on_device,
         read_from_device=True,
         greedy_only: bool = False,
+        non_greedy_decoding_on_device: bool | None = None,
```
Comment (tchedaTT) [inline models/common/warmup/warmup_utils.py:90]: This undoes changes from an earlier PR. Please fix to use the greedy_only parameter and remove the non_greedy_decoding_on_device param @dwadhwaniTT

**Review by tchedaTT — COMMENTED** [review]
This PR undoes earlier changes - probably unintentionally from a rebase, but please fix it in a followup.

File: models/common/warmup/warmup_utils.py (hunk 1 above)
Comment (dwadhwaniTT) [reply models/common/warmup/warmup_utils.py:90]: @tchedaTT Looks like this happened when I rebased, I will raise a small PR to fix this. Thanks for pointing this out!

Comment (dwadhwaniTT) [reply models/common/warmup/warmup_utils.py:90]: @tchedaTT new PR is up:

https://github.com/tenstorrent/tt-metal/pull/47248
